crates/accelerate/src/circuit_library/iqp.rs

```diff
@@ -0,0 +1,159 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::f64::consts::PI;
+
+use ndarray::{Array2, ArrayView2};
+use numpy::PyReadonlyArray2;
+use pyo3::prelude::*;
+use qiskit_circuit::{
+    circuit_data::CircuitData,
+    operations::{Param, StandardGate},
+    Qubit,
+};
+use rand::Rng;
+use smallvec::{smallvec, SmallVec};
+
+use crate::CircuitError;
+
+const PI2: f64 = PI / 2.0;
+const PI8: f64 = PI / 8.0;
+
+fn iqp(
+    interactions: ArrayView2<i64>,
+) -> impl Iterator<Item = (StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>)> + '_ {
+    let num_qubits = interactions.ncols();
+
+    // The initial and final Hadamard layer.
+    let h_layer =
+        (0..num_qubits).map(|i| (StandardGate::HGate, smallvec![], smallvec![Qubit(i as u32)]));
+
+    // The circuit interactions are powers of the CSGate, which is implemented by calling
```

Comment:
why not to call the `CSGate`, `CSdgGate` and `CZGate` instead of `CPhaseGate` ?

Reply:
I wanted to implement it the same fashion as the Python code did it, but we can also change it if you think that's better 🙂

Reply:
I agree that probably it's better to keep the same Python behavior. Do you know how these gates are used? 
I was wondering if the original code was written before we had the `CSGate & CSdgGate` in qiskit.

Reply:
Yeah the original code (4 years ago) definitely pre-dates the CS gates (2 years ago)...